contents/java/JUnit5.md

```diff
@@ -0,0 +1,243 @@
+
+```java
+public class RepeaterTest {
+  @Test
+  public void repeatTest() {
+    assertEquals(repeat("a", 5), "a\na\na\na\na\n");
+  }
+}
+```
+
+That's it! You do not have to write any additional runners or helpers to take advantage of the powerful testing features. The `@Test` tag tells JUnit 5 that the repeatTest method is meant to be a test. JUnit 5 will then automatically execute your test(s) in sequence.
+
+Let's have a look at the test results. 
+
+<img src="{{baseUrl}}/contents/java/assertFailed.PNG" />
+Hmm, the test seems to fail when we run it.
+Did you notice the subtle bug in Repeater?
+
+
+We "accidentally" caused the loop to have an extra iteration! Without the use of unit tests, we would have to manually confirm that the output indeed has 5 lines, an action that is tedious to repeat and prone to human errors.
+
+Now, let's fix the code.
+
+```java
+public class Repeater {
+  // Returns copies of reference string separated by the '\n' character.
+  public static String repeat(String reference, int copies) {
+    String result = "";
+    for (int i = 0; i < copies; i++) {
+      result += reference + "\n";
+    }
+    return result;
+  }
+}
+```
+
+<img src="{{baseUrl}}/contents/java/assertPass.PNG" />
+Running the same unit test again, we now have a passing test and we can be <trigger for="modal:somewhatAssured" trigger="click">somewhat assured</trigger> that our repeat function works as intended.
+
+### JUnit APIs
+What else can we test with JUnit? In addition to the `assertEquals` method illustrated above, JUnit allows us to test for many other scenarios. These include
+* `assertTrue`/`assertFalse`: generic checking statements for checking validity.
+* `assertNull`/`assertNotNull`: checks whether something is null or not, without causing a NullPointerException.
+* `assertArrayEquals`: yes, it loops through an entire array for you!
+* `assertThrows`/`assertDoesNotThrow`: checks that a piece of code throws the correct type of exception. In other words, JUnit even checks that the application can handle errors (or not)!
+* `assertTimeout`: ensures that a piece of code runs within time constraints - we don't want to keep the user waiting!
+
+This powerful expression means that we can write more expressive tests.
+Instead of
+```java
+try {
+  int result = divide(100, 0); //Divides 100 by 0
+} catch (ArithmeticException ae) {
+  throw new AssertionError("Divide doesn't work correctly!" + ae);
+}
+```
+We can now write
+```java
+assertThrows(ArithmeticException.class, () -> divide(100, 0));
+```
+
+This expression is a lot easier to understand, and as an added bonus, the code is also much shorter to write!
+
+### Before/After hooks
+
+Not every system can be tested so easily. Sometimes, testing is not as straightforward as running a function.
+For example, we might want to store information in an external database. We would have to first connect to the database before we can run any test.
```

Comment:
In this scenario, shouldn't the database connection be mocked? Test mocking may be outside the scope of this article, but we probably shouldn't accidentally encourage unit tests that rely on external systems. Perhaps this example can be changed to storing data in a file?

Reply:
That makes sense. I will change it to use a local file or perhaps SQLite database.